Chapter 10: Secrets (#15)
amolenk/docs: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md

We'll address referencing a secret from a Dapr component configuration later. First, we'll address calling the building block from your application code.

As with the other building blocks, the Dapr sidecar offers an API for you to interact with the Secrets Management building block. You can call this API using HTTP or gRPC. This is the URL of the API:

```http
http://localhost:<daprPort>/v1.0/secrets/<secret-store-name>/<name>?<metadata>
```

Note the Dapr specific URL segments:

- `<daprPort>` provides the port number upon which the Dapr sidecar is listening.
- `<secret-store-name>` provides the name of the selected Dapr secrets management component.
- `<name>` provides the name of the secret you want to retrieve.
- `<metadata>` provides metadata of the secret you want to retrieve. This is optional, and the available metadata properties will differ per secret store. See the [secrets management API reference]([Secrets API reference | Dapr Docs](https://docs.dapr.io/reference/api/secrets_api/)) for more information about the available metadata properties.

The JSON response you will receive contains the key and the value of the secret. Here's an example:

```http

> GET http://localhost:3500/v1.0/secrets/secrets-store/eshopsecrets?metadata.version_id=3

Reviewer: Is `metadata.version_id` store specfic or generic? I'd remove it or explain it.

```

```json
{
  "customerdb": "Server=sqlsrv-p-06;Database=CRM;User Id=sql_p_customer;Password=h8%fa51Km9D4;"
}
```

Some secret stores support storing multiple key-value pairs in a secret. In that case, you receive all the keys and values in a single JSON response like in the following example:

```http
GET http://localhost:3500/v1.0/secrets/secrets-store/interestRates
```

```json
{
  "tier1-percentage": "2.5",
  "tier2-percentage": "3.8",
  "tier3-percentage": "5.1"
}
```

### Using the Dapr .NET SDK

> Getting secrets using the Dapr .NET SDK is pretty straightforward. The DaprClient offers a method `GetSecretAsync` for retrieving secrets using the Secrets Management building block. In the following code-snippet, the connection-string for connecting to a Sql Server database is retrieved:

Reviewer: ```suggestion
Getting secrets using the Dapr .NET SDK is pretty straightforward. The `DaprClient` class offers a method `GetSecretAsync` for retrieving secrets using the secrets management building block. In the following code snippet, the connection string for connecting to a SQL Server database is retrieved:
```

```csharp

> Dictionary<string, string> metadata = new Dictionary<string, string> { { "version_id", "3" } };

Reviewer: Same comment as with the curl example: do we need to specify the version_id?

Author: No it is optional. This is more to show how to specify this metadata.

Dictionary<string,string> secrets = await daprClient.GetSecretAsync("secrets-store", "eshopsecrets", metadata);
string connectionString = secrets["customerdb"];
```

> You need to specify the name of the secret store component to use and the name of the secret you want to retrieve. The metadata is optional and differs per secret store. Notice that you always receive a dictionary as result. This is because some secret stores support storing multiple key-value pairs in a secret. In the example, we extract the secret named `customerdb` to get the connection-string.

Reviewer: ```suggestion
You need to specify the name of the secret store component to use and the name of the secret you want to retrieve. The metadata is optional and differs per secret store. Notice that you always receive a dictionary as result. This is because some secret stores support storing multiple key/value pairs in a secret. In the example, we extract the secret named `customerdb` to get the connection string.
```

> The .NET SDK also offers a .NET configuration provider for working with secrets. The provider is available in the *Dapr.Extensions.Configuration* nuget package. In the following code-snippet, you see an example of registering the secrets configuration provider in the `Program.cs` of an ASP.NET WebAPI application:  

Reviewer: ```suggestion
The .NET SDK also offers a .NET configuration provider for working with secrets. The provider is available in the *Dapr.Extensions.Configuration* NuGet package. In the following code snippet, you see an example of registering the secrets configuration provider in the `Program.cs` of an ASP.NET Web API application:  
```

```csharp
public static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
        .ConfigureAppConfiguration(config =>
        {
            var daprClient = new DaprClientBuilder().Build();
            var secretDescriptors = new List<DaprSecretDescriptor>
            {
                new DaprSecretDescriptor("eshopsecrets")
            };
            config.AddDaprSecretStore("secrets-store", secretDescriptors, daprClient);
        })
        .ConfigureWebHostDefaults(webBuilder =>
        {
            webBuilder.UseStartup<Startup>();
        });
```

> Notice how you need to specify a `DaprClient` when registering the provider. The client is used to invoke the secrets API on the Dapr sidecar of the application. You also need to specify the name of the secrets management component to use. Finally, you have to specify which secrets to retrieve by passing in a list of `DaprSecretDescriptor` instances. A `DaprSecretDescriptor` contains a secret name and optionally a dictionary containing secret metadata.

Reviewer: ```suggestion
You need to specify a `DaprClient` when registering the provider. The client is used to invoke the secrets API on the Dapr sidecar of the application. You also need to specify the name of the secrets management component to use. Finally, you have to specify which secrets to retrieve by passing in a list of `DaprSecretDescriptor` instances. A `DaprSecretDescriptor` contains a secret name and optionally a dictionary containing secret metadata.
```

```bash
kubectl create secret generic [k8s_spn_secret_name] --from-file=[pfx_certificate_file_local_path]
```

You have to specify the following command-line arguments:

- `[k8s_spn_secret_name]` is secret name in Kubernetes secret store.
- `[pfx_certificate_file_local_path]` is the path of X509 certificate file you downloaded above.

Once the Kubernetes secret is created, you can reference this in the secrets management component configuration file:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: eshop-azurekv-secret-store
  namespace: eshop
spec:
  type: secretstores.azure.keyvault
  version: v1
  metadata:
  - name: vaultName
    value: [your_keyvault_name]
  - name: spnTenantId
    value: "619926af-a7c3-4e95-93ed-4ecc4e3e652b"
  - name: spnClientId
    value: "6cf48032-6c38-43be-9d6f-2a43ce736b09"
  - name: spnCertificate
    secretKeyRef:
      name: [k8s_spn_secret_name]
      key: [pfx_certificate_file_local_name]
auth:
    secretStore: kubernetes
```

Now the eShop services running in Kubernetes can retrieve the Redis password from the Azure Key Vault.

> > It is crucial to keep the file containing the public and private key-pair for the X509 certificate of the service principal in a safe place. One way of doing this, is to make sure you place the file in a well-known folder outside the source-code repository. This well-known folder can then be used in the configuration file. On your local dev machine, you place the certificate in this folder. On the machine where the application will be deployed, the certificate is placed in this folder by an automated deployment pipeline. Also, it is a best practice to use a different service principal per environment. This ensures that the service principal used by the developers in the DEV environment cannot be used to access secrets for the PRODUCTION environment.

Reviewer: ```suggestion
> [!IMPORTANT]
> It is crucial to keep the file containing the public and private key-pair for the X509 certificate of the service principal in a safe place. One way of doing this, is to make sure you place the file in a well-known folder outside the source-code repository. This well-known folder can then be used in the configuration file. On your local dev machine, you place the certificate in this folder. On the machine where the application will be deployed, the certificate is placed in this folder by an automated deployment pipeline. Also, it is a best practice to use a different service principal per environment. This ensures that the service principal used by the developers in the DEV environment cannot be used to access secrets for the PRODUCTION environment.
```

> > It is crucial to keep the file containing the public and private key-pair for the X509 certificate of the service principal in a safe place. One way of doing this, is to make sure you place the file in a well-known folder outside the source-code repository. This well-known folder can then be used in the configuration file. On your local dev machine, you place the certificate in this folder. On the machine where the application will be deployed, the certificate is placed in this folder by an automated deployment pipeline. Also, it is a best practice to use a different service principal per environment. This ensures that the service principal used by the developers in the DEV environment cannot be used to access secrets for the PRODUCTION environment.

Reviewer: Might be good to note that if you use AKS and KV together, you can use Managed Identity

Author: Added it.

> #### Scoping secrets

Reviewer: I would place this under 'How it works'
